redis/connection.py

```diff
@@ -514,9 +516,22 @@ def __init__(
         self.host = host
         self.port = int(port)
         self.db = db
-        self.username = username
         self.client_name = client_name
+        if (username or password) and credential_provider is not None:
+            raise DataError(
+                "'username' and 'password' cannot be passed along with 'credential_"
+                "provider'. Please provide only one of the following arguments: \n"
+                "1. 'password' and (optional) 'username'\n"
+                "2. 'credential_provider'"
+            )
+
+        self.credential_provider = credential_provider
+        self.username = username
         self.password = password
+        if username or password:
+            # Keep backward compatibility by creating a static credential provider
+            # for the passed username and password
+            self.credential_provider = StaticCredentialProvider(username, password)
```

Comment:
This could break some backward compatibility in that setting `username` and `password` on an existing `Connection` won't affect the credential provider created here.

Reply:
Changed username&password to properties to address this issue

Reply:
I don't think that's a good idea - now you may end up inadvertently setting username and password on a provider that doesn't do anything with the information. Instead, I'd just do

```python
credentials = (self.credential_provider or StaticCredentialProvider(self.username, self.password)).get_credentials()
```
where you're getting the credentials...